Timeline column display order flip

manongjohn opened the pull request:
This PR changes the display order of the columns in Timeline mode so the stacking of layers is displayed bottom-up.

This puts the Timeline's layer stacking order in line with other drawing/animating software.

![image](https://user-images.githubusercontent.com/19245851/32622273-5df3e628-c550-11e7-96a3-e8ae4c596aa2.png)

### Code Changes

Xsheet Mode is based on (0,0) origin => (x,y) => (layer,frame). Columns are stacked from left-to-right by starting on the left and increasing X.

For Timeline Mode, the origin needs to be (0,bottom) => (x,y) => (frame,layer). In order to stack columns from bottom-to-top, you need to start at the bottom of the display area and decrease Y.

The changes that have been made leave the xsheet data layer as is, based on (0,0) origin, since it naturally works with array structures. This keeps changes limited to displaying of data. The changes for Timeline Mode essentially translate the (0,0) origin to the (0,bottom) origin and adjust Y accordingly.

### Timeline notes

Aside from the flipping the display of the columns in the Timeline, the following should be noted

- The infinite empty columns that follow the last real column has been removed from Timeline mode only.
	- The infinite empty columns still exists in Xsheet mode.
	- This may be added back in as a preference setting in the future.

- To add empty columns you can
	- Use the right-click context menu to "Insert" an empty column (existing option)
	- Drag columns or cells up or down, beyond the viewable area, to expose empty columns

- Home/End button now takes you to Frame# 1 and last animated Frame# respectively as it currently does in Xsheet mode

turtletooth commented:
Yes! Yes! Yes!

Agnyy commented:
Super!

gab3d commented:
niiice!! :)

Orphanlast commented:
I'm really looking forward to this change

bobbybutterfly commented:
very cool!

shun-iwasawa commented:
@manongjohn 
The current time indicator (just added in #1522 ) seems to disappear on multi column scene.
How about modifying as follows?:

```
--------------------- toonz/sources/toonz/xshrowviewer.cpp ---------------------
index b368551..df81674 100644
@@ -428,6 +428,7 @@ void RowArea::drawCurrentTimeIndicator(QPainter &p) {
   int currentRow = m_viewer->getCurrentRow();
 
   QPoint topLeft = m_viewer->positionToXY(CellPosition(currentRow, 0));
+  if (!m_viewer->orientation()->isVerticalTimeline()) topLeft.setY(0);
   QRect header   = m_viewer->orientation()
                      ->rect(PredefinedRect::FRAME_HEADER)
                      .translated(topLeft);
@@ -449,6 +450,7 @@ void RowArea::drawCurrentTimeLine(QPainter &p) {
   int currentRow = m_viewer->getCurrentRow();
 
   QPoint topLeft = m_viewer->positionToXY(CellPosition(currentRow, 0));
+  if (!m_viewer->orientation()->isVerticalTimeline()) topLeft.setY(0);
   QRect header   = m_viewer->orientation()
                      ->rect(PredefinedRect::FRAME_HEADER)
                      .translated(topLeft);
```

Thank you for adding the great improvement. I'll take a look some more.

Orphanlast commented:
@shun-iwasawa ohhh, I'm just trying to undersand, so you're also a developer. You ran into a bug, you fixed it, then shared the code with the developer assigned to this project. 

If so, that's awesome.

manongjohn commented:
@shun-iwasawa 
Ah, yes..  Funny, I was just locally testing about an hour ago to see what would conflict by merging my last 2 PRs together since I know there would be conflicts and I noticed this issue.  The hazards of working on multiple independent PRs affecting the same code.  :)

Let me update this branch with the latest OT master changes and I'll commit the corrections.

manongjohn commented:
@shun-iwasawa 
Oh, I think I just discovered the cause for the undo-able empty columns issue I noted in the description and have a correction.  I'll be fixing that issue as well before I upload my commits.

shun-iwasawa commented on toonz/sources/toonz/xsheetdragtool.cpp:
```diff
@@ -1595,22 +1597,53 @@ class ColumnMoveDragTool final : public XsheetGUI::DragTool {
+      int newEnd  = *indices.rbegin() + dCol;
+      if (newEnd > currEnd) {
+        for (int x = currEnd + 1; x <= newEnd; x++) ii.insert(x);
+        ColumnCmd::insertEmptyColumns(ii);
```
For now dragging columns or cells to upward (in timeline) will insert empty columns accordingly and there is no way to remove such empty columns other than selecting and deleting them manually.

It is open to discussion, but how about adding one (or some) of the following features?

- Dragging of cells/columns upward can insert only one empty column above for each dragging operation in timeline mode. 

- Each dragging operation will automatically remove the last empty columns on mouse release.

- Add "Remove the Last Empty Columns" command which can remove all the empty columns above the topmost occupied column.

manongjohn replied:
I'm leaning towards 2nd or 3rd option.

manongjohn commented on toonz/sources/toonz/xshcellmover.cpp:
```diff
@@ -430,32 +491,101 @@ void LevelMoverTool::onClick(const QMouseEvent *e) {
+  int dCol;
+  if (o->isVerticalTimeline()) {
+    dCol = delta.x;
+    if (origX < 0) dCol += origX;
```
Understood. I only added it to make the behavior consistent between timeline and xsheet modes, but it doesn't need to be that way. I will remove the logic that.

manongjohn commented on toonz/sources/toonz/xsheetdragtool.cpp:
```diff
@@ -1595,22 +1597,53 @@ class ColumnMoveDragTool final : public XsheetGUI::DragTool {
-    int dCol  = col - m_lastCol;
-    m_lastCol = col;
+    int dCol         = col - (m_lastCol - m_offset);
+    if (origCol < 0) dCol += origCol;
```
Per my previous comment will make sure this is removed also.

manongjohn commented:
@shun-iwasawa 
Per your suggestions, I've removed the ability to create new columns by dragging left in Xsheet mode only and added the Remove Empty Columns command which works in both modes.

Please let me know if you find anything else or if you have other ideas. Thanks!

shun-iwasawa commented on toonz/sources/toonz/xsheetdragtool.cpp:
```diff
@@ -1595,22 +1597,55 @@ class ColumnMoveDragTool final : public XsheetGUI::DragTool {
+          ColumnCmd::insertEmptyColumns(ii);
+        }
+      }
+    }
```
For now, OT crashes by doing the following operations;
1. Select multiple columns (say Col1, 2 and 3) in Xsheet view.
1. Grab a handle of Col3 and drag to the left direction.

Inserting the following lines seems to avoid crash:
```
@@ -1642,6 +1645,11 @@ public:
           ColumnCmd::insertEmptyColumns(ii);
         }
       }
+    } else {  // xsheet view
+      if (col < -m_offset) {
+        col  = -m_offset;
+        dCol = -m_lastCol;
+      }
     }
 
     if (col == (m_lastCol - m_offset)) return;
```

manongjohn replied:
Fixed.  I simply blocked the action if the 1st column number in the selection is < 0 after applying movement offset.

shun-iwasawa commented on toonz/sources/toonz/xsheetdragtool.cpp:
```diff
@@ -1595,22 +1597,55 @@ class ColumnMoveDragTool final : public XsheetGUI::DragTool {
+      if (newBegin < 0) {
+        newBegin *= -1;
+        for (int x = 0; x < newBegin; x++) ii.insert(x);
+        ColumnCmd::insertEmptyColumns(ii);
```
IMO the undo for inserting empty columns while column dragging operation should be put in a block.
How about adding the lines as follows?:
```
@@ -1600,6 +1600,9 @@ public:
     m_origOffset = m_offset = m_firstCol - col;
     assert(m_lastCol == *indices.begin());
     getViewer()->update();
+
+    if (!getViewer()->orientation()->isVerticalTimeline())
+      TUndoManager::manager()->beginBlock();
   }
   void onDrag(const CellPosition &pos) override {
     int col                     = pos.layer();

@@ -1658,14 +1666,17 @@ public:
   }
   void onRelease(const CellPosition &pos) override {
     int delta = m_lastCol - m_firstCol;
-    if (delta == 0) return;
-    TColumnSelection *selection = getViewer()->getColumnSelection();
-    std::set<int> indices       = selection->getIndices();
-    if (!indices.empty()) {
-      TUndoManager::manager()->add(new ColumnMoveUndo(indices, delta));
-      TApp::instance()->getCurrentScene()->setDirtyFlag(true);
-      TApp::instance()->getCurrentXsheet()->notifyXsheetChanged();
+    if (delta != 0) {
+      TColumnSelection *selection = getViewer()->getColumnSelection();
+      std::set<int> indices       = selection->getIndices();
+      if (!indices.empty()) {
+        TUndoManager::manager()->add(new ColumnMoveUndo(indices, delta));
+        TApp::instance()->getCurrentScene()->setDirtyFlag(true);
+        TApp::instance()->getCurrentXsheet()->notifyXsheetChanged();
+      }
     }
+    if (!getViewer()->orientation()->isVerticalTimeline())
+      TUndoManager::manager()->endBlock();
   }
 };
```

manongjohn replied:
Added. Great suggestion. Thanks

shun-iwasawa commented on toonz/sources/toonz/cellselection.cpp:
```diff
@@ -1519,6 +1543,13 @@ void TCellSelection::pasteCells() {
+      int cAdj = cellData->getColCount() - 1;
+      c0 -= cAdj;
+      c1 -= cAdj;
+    }
```
(It is not an urgent problem, but would be an issue for the future.)
You are adjusting the paste destination in order to make the cells to be pasted below the selected cell.
IMO it would better to apply the same adjustment to `Paste Into` and `Paste Numbers` commands as well, in order to gain consistency.

manongjohn replied:
I like consistency. :)  Updated to make sure all the paste options apply the timeline adjustment.

shun-iwasawa commented:
@manongjohn Sorry for the late notice, but on resolving conflicts can you please fix the following issue which I couldn't detect before merging #1573 ?

- For now, the current time indicator (the red line) is above the frame number in the row area.

It seems to be because the order of `drawCurrentTimeLine()` and `drawRows()` was swapped in `RowArea::paintEvent()`.

manongjohn commented:
@shun-iwasawa Will resolve conflicts and fix the issues you found.  Thanks.

Trying to wrap up a fix for a renumber issue I was looking into and will switch to this as soon as I can.

manongjohn commented:
@shun-iwasawa 
In addition to the change requests being done.  I also fixed the timeline indicator issue you noted as well as a few zooming adjustments that needed to be applied which affected the context menu for keyframes.

shun-iwasawa commented:
Jenkins

shun-iwasawa commented:
LGTM Thank you @manongjohn for the great improvement.
